src/parsemultipart.jl

```diff
@@ -0,0 +1,220 @@
+const CR_BYTE = 0x0d # \r
+const LF_BYTE = 0x0a # \n
+const DASH_BYTE = 0x2d # -
+const HTAB_BYTE = 0x09 # \t
+const SPACE_BYTE = 0x20
+const RETURN_BYTES = [CR_BYTE, LF_BYTE]
+
+"""
+    find_multipart_boundary(bytes, boundaryDelimiter; start::Int=1)
+
+Find the first and last index of the next boundary delimiting a part, and if
+the discovered boundary is the terminating boundary.
+"""
+function find_multipart_boundary(bytes::AbstractVector{UInt8}, boundaryDelimiter::AbstractVector{UInt8}; start::Int=1)
+    # The boundary delimiter line is prepended with two '-' characters
+    # The boundary delimiter line starts on a new line, so must be preceded by a \r\n.
+    # The boundary delimiter line ends with \r\n, and can have "optional linear whitespace" between
+    # the end of the boundary delimiter, and the \r\n.
+    # The last boundary delimiter line has an additional '--' at the end of the boundary delimiter
+    # [RFC2046 5.1.1](https://tools.ietf.org/html/rfc2046#section-5.1.1)
+
+    i = start
+    end_index = i+length(boundaryDelimiter)-1
+    while end_index <= length(bytes)
+        if bytes[i:end_index] == boundaryDelimiter
+            # boundary delimiter line start on a new line ...
+            if i > 1
+                (i == 2 || bytes[i-2] != CR_BYTE || bytes[i-1] != LF_BYTE) && error("boundary delimiter found, but it was not the start of a line")
+                # the CRLF preceding the boundary delimiter is "conceptually attached
+                # to the boundary", so account for this with the index
+                i-=2
+            end
+
+            # need to check if there are enough characters for the CRLF or for two dashes
+            end_index < length(bytes)-1 || error("boundary delimiter found, but did not end with new line")
+
+            is_terminating_delimiter = bytes[end_index+1] == DASH_BYTE && bytes[end_index+2] == DASH_BYTE
+            is_terminating_delimiter && (end_index+=2)
+
+            # ... there can be arbitrary SP and HTAB space between the boundary delimiter ...
+            while (end_index < length(bytes) && (bytes[end_index+1] in [HTAB_BYTE, SPACE_BYTE])) end_index+=1 end
+            # ... and ends with a new line
+            (end_index < length(bytes)-1) && bytes[end_index+1] == CR_BYTE && bytes[end_index+2] == LF_BYTE || error("boundary delimiter found, but did not end with new line")
+            end_index += 2
+
+            return (is_terminating_delimiter, i, end_index)
+        end
+
+        i += 1
+        end_index += 1
+    end
+
+    error("boundary delimiter not found")
+end
+
+function find_multipart_boundaries(bytes::AbstractVector{UInt8}, boundary::AbstractVector{UInt8}; start=1)
```

Comment:
Would be nice to have docstrs for all the functions in this file.

Reply:
Went through all the functions and added docstrs.  Squashed it down to the single checkin.